### PR TITLE
EM-2040 i2c fixes DO NOT MERGE

### PR DESCRIPTION
**Pull request by oliv3r**
For review purposes only!

### CODE REVIEW

File: arch/arm/boot/dts/sun7i-a20.dtsi [hunk 1]
```diff
@@ -577,16 +577,16 @@
...
 			interrupts = <GIC_SPI 58 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&ccu CLK_AHB_HDMI0>, <&ccu CLK_HDMI>,
 				 <&ccu 9>,
 				 <&ccu 18>;
 			clock-names = "ahb", "mod", "pll-0", "pll-1";
-			dmas = <&dma SUN4I_DMA_NORMAL 16>,
-			       <&dma SUN4I_DMA_NORMAL 16>,
-			       <&dma SUN4I_DMA_DEDICATED 24>;
-			dma-names = "ddc-tx", "ddc-rx", "audio-tx";
+			#clock-cells = <0>;
```
Comment (MarcoTvM) [inline arch/arm/boot/dts/sun7i-a20.dtsi:586]: Commented out does not need to be checked in

File: arch/arm/boot/dts/sun7i-a20.dtsi [hunk 2]
```diff
@@ -617,6 +617,15 @@
 			};
 		};
 
+		hdmi_i2c: i2c@016500 {
+			compatible = "allwinner,sun7i-a20-hdmi-i2c";
+			reg = <0x01c16500 0x40>;
+			#address-cells = <1>;
+			#size-cells = <0>;
```
Comment (MarcoTvM) [inline arch/arm/boot/dts/sun7i-a20.dtsi:624]: Commented out does not need to be checked in

File: arch/arm/boot/dts/sun7i-a20.dtsi [hunk 3]
```diff
@@ -617,6 +617,15 @@
 			};
 		};
 
+		hdmi_i2c: i2c@016500 {
+			compatible = "allwinner,sun7i-a20-hdmi-i2c";
+			reg = <0x01c16500 0x40>;
+			#address-cells = <1>;
```
Comment (MarcoTvM) [inline arch/arm/boot/dts/sun7i-a20.dtsi:623]: Commented out does not need to be checked in

File: drivers/gpu/drm/sun4i/sun4i_hdmi_ddc_clk.c [hunk 4]
```diff
@@ -109,38 +116,34 @@ static const struct clk_ops sun4i_ddc_ops = {
...
 
-int sun4i_ddc_create(struct sun4i_hdmi *hdmi, struct clk *parent)
+struct clk *sun4i_ddc_create(struct device *dev, struct regmap *regmap,
+			     const struct sun4i_hdmi_i2c_variant *variant,
+			     const struct clk *parent_clk)
 {
 	struct clk_init_data init;
 	struct sun4i_ddc *ddc;
-	const char *parent_name;
+	const char *parents[SUN4I_HDMI_I2C_CLK_NUM_PARENTS];
```
Comment (MarcoTvM) [inline drivers/gpu/drm/sun4i/sun4i_hdmi_ddc_clk.c:125]: This creates a local (on stack) array of 1 character (define is 1), yet, this is being passed on indirectly as a pointer to other places using structures. See comment further on.

File: drivers/gpu/drm/sun4i/sun4i_hdmi_ddc_clk.c [hunk 5]
```diff
@@ -109,38 +116,34 @@ static const struct clk_ops sun4i_ddc_ops = {
...
 
-	parent_name = __clk_get_name(parent);
-	if (!parent_name)
-		return -ENODEV;
+	parents[0] = __clk_get_name(parent_clk);
```
Comment (MarcoTvM) [inline drivers/gpu/drm/sun4i/sun4i_hdmi_ddc_clk.c:127]: Filling the actual value here, in original code, you get a char pointer

File: drivers/gpu/drm/sun4i/sun4i_hdmi_ddc_clk.c [hunk 6]
```diff
@@ -109,38 +116,34 @@ static const struct clk_ops sun4i_ddc_ops = {
...
 	if (IS_ERR(ddc->reg))
-		return PTR_ERR(ddc->reg);
+		return ERR_CAST(ddc->reg);
 
-	init.name = "hdmi-ddc";
+	init.name = "hdmi-i2c";
 	init.ops = &sun4i_ddc_ops;
-	init.parent_names = &parent_name;
-	init.num_parents = 1;
+	init.parent_names = parents;
```
Comment (MarcoTvM) [inline drivers/gpu/drm/sun4i/sun4i_hdmi_ddc_clk.c:141]: This should be &parents[0]?
Should parents be made static?

File: drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+	 *
+	 * If threshold is inclusive, we can set the TX threshold to
+	 * 0 instead of 1.
+	 */
+	regmap_field_write(drv->field_ddc_fifo_tx_thres,
+			   drv->variant->ddc_fifo_thres_incl ? 0 : 1);
+	regmap_field_write(drv->field_ddc_fifo_rx_thres, RX_THRESHOLD);
+	regmap_field_write(drv->field_ddc_fifo_clear, 1);
+	if (regmap_field_read_poll_timeout(drv->field_ddc_fifo_clear,
+					   reg, !reg, 100, 2000))
```
Comment (MarcoTvM) [inline drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c:200]: introduce defines that describe what 100 and 2000 mean?
seconds? milisecond? usec?

File: drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+	/*
+	 * Limit transfer length by FIFO threshold or FIFO size.
+	 * For TX the threshold is for an empty FIFO.
+	 */
+	len = min_t(int, len, read ? read_len : SUN4I_HDMI_DDC_FIFO_SIZE);
+
+	/* Wait until error, FIFO request bit set or transfer complete */
+	if (regmap_field_read_poll_timeout(drv->field_ddc_int_status, reg,
+					   reg & mask, len * byte_time_ns,
+					   100000))
```
Comment (MarcoTvM) [inline drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c:150]: introduce define to make clear what this magical number means

File: drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+	for (i = 0; i < msg->len; i += len) {
+		len = fifo_transfer(drv, msg->buf + i, msg->len - i,
+				    msg->flags & I2C_M_RD);
+		if (len <= 0)
+			return len;
+	}
+
+	/* Wait for command to finish */
+	if (regmap_field_read_poll_timeout(drv->field_ddc_start,
+					   reg, !reg, 100, 100000))
```
Comment (MarcoTvM) [inline drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c:231]: More of the same, magical values!

**Review by MarcoTvM — CHANGES_REQUESTED** [review]
Mostly magical numbers, but one issue with pointers that worries me.

**Comment by oliv3r** [comment]
New PR #4

**Pull request closed without merging**